### PR TITLE
Fixup leveling

### PR DESCRIPTION
**Pull request by thinkyhead**
- Init `zprobe_zoffset`
- Remove `current_position[Z_AXIS] = zprobe_zoffset` lines from the
  `set_bed_level_equation_*` functions
- Apply standards to `mesh_bed_leveling` files
- Document `MESH_BED_LEVELING`
- Make `zprobe_zoffset` conditional
- Fix ConfigurationStore for `zprobe_zoffset`

### CODE REVIEW

File: Marlin/Marlin_main.cpp [hunk 1]
```diff
@@ -1095,10 +1095,7 @@ static void set_bed_level_equation_lsq(double *plane_equation_coefficients)
 //    corrected_position.debug("position after");
     current_position[X_AXIS] = corrected_position.x;
     current_position[Y_AXIS] = corrected_position.y;
-    current_position[Z_AXIS] = corrected_position.z;
-
-    // put the bed at 0 so we don't go below it.
-    current_position[Z_AXIS] = zprobe_zoffset; // in the lsq we reach here after raising the extruder due to the loop structure
+    current_position[Z_AXIS] = corrected_position.z + zprobe_zoffset;
```
Comment (thinkyhead) [inline Marlin/Marlin_main.cpp]: Still waiting to find out whether this is good or bad. @nophead leaning towards bad…

**Comment by thinkyhead** [comment]
Failing Travis CI due to problems at their end.

**Comment by nophead** [comment]
This comment make me think adding the probe offset is to compensate for the probing code somehow messing up the z position.

```
in the lsq we reach here after raising the extruder due to the loop structure
```

I am beginning to think perhaps Marlin's ABL only compensates for the tilt and not for the offset. Can this be true or am I missing something?

**Comment by harerama** [comment]
as said in https://github.com/MarlinFirmware/Marlin/issues/1682, I think `zprobe_zoffset` should not be added here.

**Comment by tonnico** [comment]
Maybe this is also the reason why you must G28 Z before probing?!?

**Comment by thinkyhead** [comment]
I've taken all the feedback and produced a final PR that addresses all the recent issues in leveling. This also removes a few compiler warnings visible when using the -Wall flag.

**Comment by thinkyhead** [comment]
@nophead I also notice the old comment above that old line:

```
// put the bed at 0 so we don't go below it.
```

... But again, I don't know what that really means here. Maybe it used to set Z to 0. I'll have to go back and look at old sources soon.

**Comment by alexborro** [comment]
@nophead,  you are correct. The current code just compensate the tilt. The
offset must be adjusted by Z_Probe_Offset.
But once set a working Z_Probe_Offset, you can raise or lower the bed and
the offset should be right.
Em 27/03/2015 05:48, "Chris" notifications@github.com escreveu:

**Comment by thinkyhead** [comment]
So, just to get the final consensus, this…

```
current_position[Z_AXIS] = corrected_position.z;
```

Not this…

```
current_position[Z_AXIS] = corrected_position.z + zprobe_zoffset;
```

And not this either…

```
current_position[Z_AXIS] = corrected_position.z - zprobe_zoffset;
```

**Comment by alexborro** [comment]
@thinkyhead, check the other topic where a tester said the current Marlin
version is closer to the physical probe offset.. He had to double the
offset to correct the Z position after your last merge.
Em 27/03/2015 21:08, "Scott Lahteine" notifications@github.com escreveu:

**Comment by thinkyhead** [comment]
@alexborro Yeah, I saw that…. I'd like to see what the `current_position` values are throughout the process, and where this is getting thrown off. And then document the sh*t out of the code.

**Pull request merged**